File: 1-array-question/matrix/spiralOrder/a1.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function (matrix) {
  let r = matrix.length,
    c = matrix[0].length
  let [startRow, endRow, startCol, endCol] = [0, r - 1, 0, c - 1]

  let arr = []

  while (startRow <= endCol && startCol <= endRow) {
    for (let i = startRow; i <= endCol; i++) arr.push(matrix[startCol][i])
    startCol++

    for (let i = startCol; i <= endRow; i++) arr.push(matrix[i][endCol])
    endCol--

    if (startCol <= endRow) {
      // condition 1
      for (let i = endCol; i >= startRow; i--) arr.push(matrix[endRow][i])
      endRow--
    }

    if (startRow <= endCol) {
      // condition 2
      for (let i = endRow; i >= startCol; i--) arr.push(matrix[i][startRow])
      startRow++
    }
  }

  return arr
}

File: 2-array/1-array-question/matrix/spiralOrder/a1.cpp
// class Solution {
// public:
//     vector<int> spiralOrder(vector<vector<int>>& matrix) {
//         vector<int> ans;
//         int count = 0 ;
//         int m = matrix.size();
//         int n = matrix[0].size();
//         int endRow = m - 1;
//         int endCol = n - 1;
//         int startRow = 0;
//         int startCol = 0;
//         int totalElement = m * n ;

//         while(count < totalElement){
//             // col is changing --> so i will varry acc to it
//             //^ wrote i = 0 fix assuming
//             for(int i = startCol ; i <=  endCol && count < totalElement ; i++){
//                 ans.push_back(matrix[startRow][i]);
//                 count++;
//             }
//             startRow++;
//             //print lastCol ::  row changing , col same
//             for(int i = startRow ; i <= endRow && count < totalElement ; i++){
//                 ans.push_back(matrix[i][endCol]);
//                 count++;
//             }
//             endCol--;
//             //print lastRow ::  row same , col changing
//             //! have limit of startCol not startRow
//             for(int i = endCol ; i >= startCol && count < totalElement ; i--){
//                 ans.push_back(matrix[endRow][i]);
//                 count++;
//             }
//             endRow--;
//             //^  consider equal , allready ++ 
//             for(int i = endRow ; i >= startRow && count < totalElement ; i--){
//                 ans.push_back(matrix[i][startCol]);
//                 count++;
//             }
//             startCol++;
//         }
//         return ans;
//     }
// };